### PR TITLE
CI Infra: Add terraform code to setup static Infra

### PR DESCRIPTION
**Pull request by surajssd**
The code in this PR will be used to set up static infrastructure which will be consumed by the the test provisioner for Azure provider.

### CODE REVIEW

**Comment by surajssd** [comment]
For testing:

## Initial Setup

Create the needed resource group, storage account and container for the terraform state.

```bash
export RG="REPLACE"
export LOCATION="eastus"
export SA_NAME="REPLACE"
export CTR="REPLACE"
```

```bash
az group create \
    --name $RG \
    --location $LOCATION

az storage account create \
    --name $SA_NAME \
    --resource-group $RG \
    --location $LOCATION \
    --sku Standard_LRS \
    --kind StorageV2

az storage container create \
    --name CTR \
    --account-name $SA_NAME
```

Once done, as per the above values, in `providers.tf` file, replace the values in `backend "azurerm" {}` block.

## Terraform

Login using `az login` and run the following commands:

```
terraform init
terraform apply
```

Once the above exits successfully, make a note of the above env var output.

## GH side set up

- Create a github repository with the name equal to the value you provided to the variable `gh_repo` when running `terraform apply`.
- Once the repository is created head over to the repository's `Settings > Security (side bar) > Secrets and Variables > Actions > Secrets (tab) > New repository secret` and create one secret at a time.
- Now clone this repository locally and run the following commands:

```bash
mkdir -p .github/workflows
echo "
name: Run Azure Login with OIDC
on: [push]

permissions:
      id-token: write
      contents: read
jobs:
  build-and-deploy:
    runs-on: ubuntu-latest
    steps:
      - name: 'Az CLI login'
        uses: azure/login@v1
        with:
          client-id: ${{ secrets.AZURE_CLIENT_ID }}
          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}

      - name: 'Run az commands'
        run: |
          az account show
          az group list
" | tee .github/workflows/main.yml
```

- Commit and push this change and head over to the repository's `Actions` tab to see the GH run in action. Now you would see that the run has successfully happened.

---

PS: We don't need to publish ACR secrets either, even the push can happen automatically we just need to sign in to the registry by running: `az acr login --name <registry name>`.

File: ci-infra/azure/main.tf
Comment (surajssd) [inline ci-infra/azure/main.tf:33]: I have to see what subject we can provide for pull request specific code change.

**Comment by surajssd** [comment]
@bpradipt @wainersm @mkulke @kartikjoshi21 can you take a look and tell me if this looks okay? I will add documentation after this.

File: ci-infra/azure/variables.tf
Comment (mkulke) [inline ci-infra/azure/variables.tf:1]: What is the reasoning for this variable?

File: ci-infra/azure/outputs.tf
Comment (mkulke) [inline ci-infra/azure/outputs.tf:5]:
```suggestion

```

File: ci-infra/azure/variables.tf
Comment (surajssd) [reply ci-infra/azure/variables.tf:1]: I added `ver` so that it becomes easier to switch, if we needed to. From one set up to a new one. The problem with having consistent naming shows up in recreating storage accounts, container registries, etc.

Comment (surajssd) [reply ci-infra/azure/variables.tf:1]: Is there a better way to tackle these problems?

Comment (mkulke) [reply ci-infra/azure/variables.tf:1]: I'm not sure I understood the problem yet, you intend to keep multiple versions of a terraform deployment? If that's the case, wouldn't bumping the ver variable just recreate the resources with new names?

Comment (mkulke) [reply ci-infra/azure/variables.tf:1]: and if there are name clashes with non-tf managed resources does `random_id()` help?

Comment (surajssd) [reply ci-infra/azure/variables.tf:1]: > you intend to keep multiple versions of a terraform deployment?

Yes for a short while before tearing down old set up and still want to recreate a new set up.

**Comment by wainersm** [comment]
Hi @surajssd , that's a great idea! I'm starting to review this PR but as I am not used to this setup I might be a little slow.

File: ci-infra/azure/variables.tf
Comment (mkulke) [reply ci-infra/azure/variables.tf:1]: So that's not what you want, right? A `tf apply` would remove the old resources and create the new ones in one go. If I understand correctly you want to have some kind of blue/green deployment?

File: ci-infra/azure/main.tf
Comment (wainersm) [reply ci-infra/azure/main.tf:33]: Are you thinking in secure it further, like, allowing only jobs triggered by `actor` that belong to the committers group?  Or perhaps we could add additional checkers on the jobs themself, for example, allow to run the e2e jobs only the actors belonging the commiters group.

One interesting thing is that when you have a secret hard-coded on the repository, it is not visible to jobs running on `pull_request` events. @fidencio and I learned it the hard way when setting the workflows to kata containers, which currently they run on `pull_request_target` events (secrets are exposed to `pull_request_target`). I assume that with GitHub's OIDC provider the secret will be exposed to `pull_request` events too.

**Comment by wainersm** [comment]
@surajssd I still didn't understand the magic to make the workflow/job to request the token via Github's OIDC provider. I read this https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc but it seems to not apply.

**Comment by mkulke** [comment]
> @surajssd I still didn't understand the magic to make the workflow/job to request the token via Github's OIDC provider. I read this https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc but it seems to not apply.

The steps are described in this doc:

https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-azure

File: ci-infra/azure/variables.tf
Comment (surajssd) [reply ci-infra/azure/variables.tf:1]: Yes, you are right. Hmm, makes me wonder probably we don't need this everywhere. Only for things that will conflict, like storage account, registry name, etc.

Comment (surajssd) [reply ci-infra/azure/variables.tf:1]: I think even with clean up it is fine. As in we will have a whole new set up every time we run it. Especially in environments where things will require no sharing with previous setups.

**Comment by surajssd** [comment]
Added documentation as well. Once merged, I will apply this on our CI account.

**Comment by bpradipt** [comment]
@surajssd the approach looks fine to me.

File: ci-infra/azure/variables.tf
Comment (bpradipt) [inline ci-infra/azure/variables.tf:4]: I have mostly seen the usage of variable name `id` for such scenarios.  Seems a matter of preferences.

**Comment by surajssd** [comment]
@bpradipt feel free to merge it ;-)

**Comment by kartikjoshi21** [comment]
LGTM as well, once merged i will use this to test my ongoing CI work. Thanks @surajssd

**Review by kartikjoshi21 — APPROVED** [review]

**Comment by kartikjoshi21** [comment]
@mkulke @bpradipt can you please look at it again and merge if looks good ? So that i can start testing my work. Thanks

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @surajssd

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @surajssd

**Pull request merged**
